### PR TITLE
Trim common indent when rendering snippets in docs

### PR DESCRIPTION
**Pull request by alllex**
Build DSL snippets used in the user manual often use only parts of the actual testable build script samples. Due the DSL differences between Groovy and Kotlin, sometimes the snippets selected to be displayed in the docs have different indentation. This may be due to Kotlin requiring an extra context being explicitly introduced. Some snippets actually select indented code from the sample project.

The problem is that these implementation details leak into the manual intended for the user and the indentation does not make any sense in the context of the manual.

This PR strips any common indentation (max number of whitespace characters leading each line) of code snippets at the stage of embedding them into the rendered docs.

Before:
image

After:
image

Fixes issues described in:
- https://github.com/gradle/gradle/pull/23105

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test BD

File: subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java [hunk 1]
```diff
@@ -181,4 +182,37 @@ private String determineActiveTag(String line, List<String> tags) {
...
+            }
+
+            int indent = 0;
+            while (indent < line.length() && Character.isWhitespace(line.charAt(indent))) {
+                indent++;
+            }
+            if (indent < line.length()) {
+                minIndent = Math.min(minIndent, indent);
+            }
+        }
```
Comment (reinsch82) [inline subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java:201]: could you put this in a method?
e.g. `calcMinIntent`

File: subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java [hunk 2]
```diff
@@ -181,4 +182,37 @@ private String determineActiveTag(String line, List<String> tags) {
         }
         return null;
     }
+
+    private static String trimIndent(String source) {
+        String[] lines = source.split("\r\n|\r|\n");
+        int minIndent = Integer.MAX_VALUE;
+        for (String line : lines) {
+            if (line.trim().length() == 0) {
```
Comment (reinsch82) [inline subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java:190]: `isEmpty()` ?

File: subprojects/docs-asciidoctor-extensions-base/src/test/groovy/org/gradle/docs/asciidoctor/SampleIncludeProcessorTest.groovy [hunk 3]
```diff
@@ -223,4 +223,54 @@ include::sample[dir="src/samples",files="foo.xml[tag=bar]"]
...
+                // Comment
+                doLast {
+    // end::foo[]
+                    println "hello world"
+    // tag::foo[]
+                }
+    // end::foo[]
+"""
+
+        String asciidocContent = """
```
Comment (jbartok) [inline subprojects/docs-asciidoctor-extensions-base/src/test/groovy/org/gradle/docs/asciidoctor/SampleIncludeProcessorTest.groovy:263]: I don't like tests formatted like this. Could you maybe use the stripMargin style instead?
```
                         ```ABC
                            |123
                            |456```.stripMargin()
```

**Review by jbartok — CHANGES_REQUESTED** [review]
Looks good, minus the few comments inlined.

**Review by reinsch82 — CHANGES_REQUESTED** [review]
2 minor requests :)

File: subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java (hunk 1 above)
Comment (alllex) [reply subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java]: Done

File: subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java (hunk 2 above)
Comment (alllex) [reply subprojects/docs-asciidoctor-extensions-base/src/main/java/org/gradle/docs/asciidoctor/SampleIncludeProcessor.java]: Done

File: subprojects/docs-asciidoctor-extensions-base/src/test/groovy/org/gradle/docs/asciidoctor/SampleIncludeProcessorTest.groovy (hunk 3 above)
Comment (alllex) [reply subprojects/docs-asciidoctor-extensions-base/src/test/groovy/org/gradle/docs/asciidoctor/SampleIncludeProcessorTest.groovy:271]: I just found that combining `.trim().stripMargin()` gives the best of both worlds:
- `trim()` allows to keep `"""` quotes on separate lines not making the first and the last line of multi-line string stand out
- `stripMargin()` allows to still use `|` margins to make it clear how much whitespace we want for each line

The result is the test code is much easier to follow visually once you recognized `trim().stripMargin()` pattern:
```
        def expectedContent = '''
            |// Comment
            |doLast {
            |    println "hello world"
            |}
        '''.trim().stripMargin()
```

Comment (jbartok) [reply subprojects/docs-asciidoctor-extensions-base/src/test/groovy/org/gradle/docs/asciidoctor/SampleIncludeProcessorTest.groovy:271]: Yeah, waay better.

**Review by jbartok — APPROVED** [review]

**Review by reinsch82 — APPROVED** [review]
LGTM

**Review by hythloda — APPROVED** [review]
elegant soution

**Comment by alllex** [comment]
@bot-gradle test and merge

**Pull request merged**